File: src/android/app/src/main/java/com/dogwalking/app/domain/usecases/PaymentUseCase.kt
package com.dogwalking.app.domain.usecases

// -------------------------------------------------------------------------------------------------
// External Imports with Specified Versions
// -------------------------------------------------------------------------------------------------
import javax.inject.Inject // version 1
import kotlinx.coroutines.flow.Flow // version 1.7.3
import kotlinx.coroutines.flow.flow // version 1.7.3
import timber.log.Timber // version 5.0.1

// -------------------------------------------------------------------------------------------------
// Internal Imports (Named)
// -------------------------------------------------------------------------------------------------
import com.dogwalking.app.data.repository.PaymentRepository
import com.dogwalking.app.domain.models.Payment

/**
 * Placeholder class representing analytics functionality for payment operations.
 * Provides methods to track custom metrics, usage events, and success/failure rates.
 */
class PaymentAnalytics {
    /**
     * Tracks any general event with an optional key-value metadata map.
     */
    fun trackEvent(eventName: String, metadata: Map<String, Any>? = null) {
        // Implementation for analytics service integration goes here.
        Timber.d("Analytics Event: $eventName, Metadata: $metadata")
    }

    /**
     * Tracks payment-specific metrics such as completion rates, failures, or refund statuses.
     */
    fun trackPaymentMetric(metricName: String, payment: Payment) {
        // Implementation for payment-specific analytics goes here.
        Timber.d("Analytics Metric: $metricName, Payment ID: ${payment.id}, Status: ${payment.status}")
    }
}

/**
 * Placeholder class representing a validator for payment data. It includes methods for checking
 * compliance, fraud detection, or any advanced business rules prior to actual processing.
 */
class PaymentValidator {
    /**
     * Validates whether the payment is structurally correct (amount > 0, necessary fields, etc.).
     * This method might also handle 3D Secure checks, card validation, or other advanced tutorials.
     */
    fun isValidPayment(payment: Payment): Boolean {
        // Placeholder example: Payment amount must be > 0
        return payment.amount > 0.0
    }

    /**
     * Runs additional fraud checks or risk scoring. Returns true if the payment is considered safe.
     */
    fun isFraudCheckPassed(payment: Payment): Boolean {
        // Example placeholder logic
        return true
    }
}

/**
 * Placeholder data class representing the structure of a payment filter, including any query
 * parameters such as date ranges or status filters. In a production scenario, this could be
 * more comprehensive.
 */
data class PaymentFilter(
    val status: String? = null,
    val minAmount: Double? = null,
    val maxAmount: Double? = null
)

/**
 * Placeholder data class for pagination configuration, such as page number and page size.
 */
data class Pagination(
    val page: Int,
    val pageSize: Int
)

/**
 * Placeholder data class representing a generic paginated list result that wraps the data
 * and includes metadata such as total items and page count.
 */
data class PaginatedList<T>(
    val items: List<T>,
    val totalItems: Int,
    val currentPage: Int,
    val totalPages: Int
)

/**
 * Placeholder data class for a validation result. Typically, this would contain flags or
 * error messages detailing any issues with the payment data.
 */
data class ValidationResult(
    val isValid: Boolean,
    val errorMessage: String? = null
)

/**
 * Placeholder data class representing a refund. In a real system, this might hold the refund
 * amount, identifiers from the payment gateway, or any dispute/resolution info.
 */
data class Refund(
    val refundId: String,
    val amount: Double,
    val status: String
)

/**
 * Placeholder data class for a refund request. The fields may vary depending on your
 * payment gateway integration.
 */
data class RefundRequest(
    val amount: Double,
    val reason: String
)

/**
 * Enhanced use case class implementing comprehensive payment business logic with security,
 * validation, and offline support. This class aligns with the system's financial operations
 * requirements under scope (1.3), payment processing design (2.1), and offline-first
 * architecture (2.2.1).
 */
class PaymentUseCase @Inject constructor(
    private val paymentRepository: PaymentRepository,  // Step 1: Initialize payment repository dependency
    private val analytics: PaymentAnalytics,           // Step 2: Initialize analytics tracker
    private val validator: PaymentValidator            // Step 3: Initialize payment validator
) {

    init {
        // Step 4: Setup error handlers or any additional initialization.
        // For example, we might configure default exception handlers or log certain states.
        Timber.d("PaymentUseCase initialized with PaymentRepository, PaymentAnalytics, and PaymentValidator.")
    }

    // ---------------------------------------------------------------------------------------------
    // FUNCTION: processPayment
    // Description: Processes a new payment with enhanced validation and offline support. Returns a
    // Flow<Result<Payment>> to accommodate asynchronous operation and possible failures. Fulfills
    // the requirement for secure payment, error handling, and potential offline queueing described
    // in the specification.
    // ---------------------------------------------------------------------------------------------
    fun processPayment(payment: Payment): Flow<Result<Payment>> = flow {
        // STEP 1: Log payment attempt
        Timber.i("Starting processPayment for Payment ID: ${payment.id}")

        // STEP 2: Validate payment details using validator
        if (!validator.isValidPayment(payment) || !validator.isFraudCheckPassed(payment)) {
            val errorMsg = "Payment validation failed for Payment ID: ${payment.id}"
            Timber.e(errorMsg)
            emit(Result.failure(Exception(errorMsg)))
            return@flow
        }

        // STEP 3: Check network connectivity (Placeholder: simulate isNetworkAvailable)
        val isNetworkAvailable = isNetworkAvailable()
        Timber.d("Network connectivity check for Payment ID: ${payment.id}: $isNetworkAvailable")

        // STEP 4: If offline, queue payment for later processing
        if (!isNetworkAvailable) {
            try {
                paymentRepository.queueOfflinePayment(payment)
                analytics.trackPaymentMetric("payment_queued_offline", payment)
                Timber.w("Payment queued offline due to no connectivity, Payment ID: ${payment.id}")
                emit(Result.success(payment.copy(status = payment.status))) // Return current payment info
            } catch (ex: Exception) {
                Timber.e(ex, "Failed to queue payment offline for Payment ID: ${payment.id}")
                emit(Result.failure(ex))
            }
            return@flow
        }

        // STEP 5: If online, process payment through gateway (simulated by repository.processPayment)
        try {
            val result = paymentRepository.processPayment(payment)
            // The repository might return a Payment with updated status or throw exceptions on error.
            analytics.trackPaymentMetric("payment_gateway_attempt", payment)
            emit(Result.success(result))
        } catch (ex: Exception) {
            // STEP 7: Handle errors with retry mechanism if needed or direct failure
            Timber.e(ex, "Error processing online payment for Payment ID: ${payment.id}")
            emit(Result.failure(ex))
        }

        // STEP 8: In a real scenario, we might actively update the Payment status in the repository.
        // For demonstration, we rely on the internal "processPayment" to do so, or we could:
        // paymentRepository.updatePaymentStatus(payment.id, PaymentStatus.COMPLETED)
        // ensuring the final status is reflected in local/offline storage.

        // STEP 9: Return processing result. The .emit() calls above have already covered success/failure.
    }

    // ---------------------------------------------------------------------------------------------
    // FUNCTION: getPaymentHistory
    // Description: Retrieves filtered payment history with pagination. Returns a Flow<PaginatedList<Payment>>
    // to handle asynchronous data streams and offline caching. This aligns with the project goal of
    // providing secure transaction history access.
    // ---------------------------------------------------------------------------------------------
    fun getPaymentHistory(
        userId: String,
        filter: PaymentFilter,
        pagination: Pagination
    ): Flow<PaginatedList<Payment>> = flow {
        // STEP 1: Validate user ID
        require(userId.isNotBlank()) {
            "User ID cannot be blank when fetching payment history."
        }

        Timber.d("Fetching payment history for user $userId with filter=$filter, pagination=$pagination")

        // STEP 2: Apply payment filters - in a robust system, you'd pass the filter into a repository.
        // For demonstration, we highlight placeholder usage. The repository might accept these directly.

        // STEP 3: Retrieve paginated history from repository
        // Placeholder for an actual repository call (the PaymentRepository might require specialized methods):
        val fullPayments = paymentRepository.getPaymentHistory(userId) // might directly retrieve raw list

        // STEP 4: Apply security checks or additional filtering logic
        // For demonstration, we simply filter by status if provided
        val filteredPayments = if (filter.status != null) {
            fullPayments.filter { it.status.name == filter.status }
        } else {
            fullPayments
        }.filter {
            (filter.minAmount == null || it.amount >= filter.minAmount) &&
            (filter.maxAmount == null || it.amount <= filter.maxAmount)
        }

        // Implement pagination
        val startIndex = (pagination.page - 1) * pagination.pageSize
        val endIndex = (startIndex + pagination.pageSize).coerceAtMost(filteredPayments.size)
        val pagedItems = if (startIndex in filteredPayments.indices) {
            filteredPayments.subList(startIndex, endIndex)
        } else {
            emptyList()
        }

        val totalItems = filteredPayments.size
        val totalPages = if (pagination.pageSize > 0) {
            (totalItems + pagination.pageSize - 1) / pagination.pageSize
        } else {
            1
        }

        // Construct the paginated result
        val paginatedList = PaginatedList(
            items = pagedItems,
            totalItems = totalItems,
            currentPage = pagination.page,
            totalPages = totalPages
        )

        // STEP 5: Track analytics
        analytics.trackEvent("get_payment_history", mapOf("userId" to userId, "page" to pagination.page))

        // STEP 6: Return filtered results
        emit(paginatedList)
    }

    // ---------------------------------------------------------------------------------------------
    // FUNCTION: validatePayment
    // Description: Comprehensive payment validation with fraud detection. Returns a Result<ValidationResult>.
    // This method encapsulates deeper checks for compliance, user eligibility, and potential risk.
    // ---------------------------------------------------------------------------------------------
    fun validatePayment(payment: Payment): Result<ValidationResult> {
        Timber.d("Initiating validation for Payment ID: ${payment.id}")

        // STEP 1: Validate amount and currency (placeholder: Payment domain might have a currency field)
        if (payment.amount <= 0) {
            return Result.success(
                ValidationResult(
                    isValid = false,
                    errorMessage = "Payment amount must be greater than zero."
                )
            )
        }

        // STEP 2: Check payment method validity (placeholder)
        // Real logic would ensure a recognized payment method or card brand is used.

        // STEP 3: Verify user eligibility (placeholder: might integrate with user domain checks).

        // STEP 4: Run fraud detection rules
        if (!validator.isFraudCheckPassed(payment)) {
            return Result.success(
                ValidationResult(
                    isValid = false,
                    errorMessage = "Fraud check failed."
                )
            )
        }

        // STEP 5: Validate geographic restrictions (placeholder).

        // STEP 6: Check rate limits (placeholder).

        // If everything passes, return a valid result
        return Result.success(
            ValidationResult(
                isValid = true,
                errorMessage = null
            )
        )
    }

    // ---------------------------------------------------------------------------------------------
    // FUNCTION: refundPayment
    // Description: Processes full or partial payment refund, returning a Flow<Result<Refund>>. Detailed
    // logic includes verifying eligibility, updating statuses, and tracking analytics.
    // ---------------------------------------------------------------------------------------------
    fun refundPayment(paymentId: String, request: RefundRequest): Flow<Result<Refund>> = flow {
        Timber.i("Refund initiated for Payment ID: $paymentId with request: $request")

        // STEP 1: Validate refund eligibility. In a real scenario, we might look up payment info,
        // check if it's refundable, etc.
        val targetPayment = try {
            paymentRepository.getPaymentHistory("") // placeholder to get all payments
                .firstOrNull { it.id == paymentId }
                ?: throw IllegalArgumentException("Payment with ID $paymentId not found.")
        } catch (ex: Exception) {
            emit(Result.failure(ex))
            return@flow
        }

        if (!targetPayment.isRefundable) {
            val msg = "Payment with ID $paymentId is not eligible for refund."
            Timber.e(msg)
            emit(Result.failure(Exception(msg)))
            return@flow
        }

        // STEP 2: Calculate refund amount. In some cases, partial refunds are allowed.
        val refundAmount = request.amount
        if (refundAmount <= 0 || refundAmount > targetPayment.amount) {
            val msg = "Invalid refund amount: $refundAmount for Payment ID: $paymentId"
            Timber.e(msg)
            emit(Result.failure(Exception(msg)))
            return@flow
        }

        // STEP 3: Process refund through gateway (Placeholder: simulate success)
        val refundId = "refund_${paymentId}"
        val newRefund = Refund(
            refundId = refundId,
            amount = refundAmount,
            status = "SUCCESS"
        )

        // STEP 4: Update payment status in a real scenario. PaymentRepository might have
        // a method to handle partial or full refunds. We'll simulate a partial update:
        try {
            paymentRepository.updatePaymentStatus(
                paymentId = targetPayment.id,
                newStatus = "REFUNDED",  // or "PARTIALLY_REFUNDED"
                synced = false,
                updatedTimestamp = System.currentTimeMillis()
            )
        } catch (ex: Exception) {
            Timber.e(ex, "Error updating payment status during refund, Payment ID: $paymentId")
            emit(Result.failure(ex))
            return@flow
        }

        // STEP 5: Track refund metrics
        analytics.trackPaymentMetric("refund_processed", targetPayment)

        // STEP 6: Return refund result
        emit(Result.success(newRefund))
    }

    // ---------------------------------------------------------------------------------------------
    // FUNCTION: syncOfflinePayments
    // Description: Synchronizes queued offline payments, returning a Flow<List<Result<Payment>>>
    // describing the outcome of each synchronization attempt. This meets the offline-first requirement.
    // ---------------------------------------------------------------------------------------------
    fun syncOfflinePayments(): Flow<List<Result<Payment>>> = flow {
        Timber.i("Initiating offline payment synchronization")

        val results = mutableListOf<Result<Payment>>()

        // STEP 1: Check network connectivity
        if (!isNetworkAvailable()) {
            // If offline, do nothing
            Timber.w("Network is unavailable; offline payments cannot be synced at this time.")
            emit(results)
            return@flow
        }

        // STEP 2: Retrieve queued payments
        val queuedPayments = try {
            paymentRepository.syncOfflinePayments() // If there's a method that returns offline payments
            // The above might be a synchronous call or a flow in the repository. Adjust accordingly.
            // We'll assume it's synchronous for demonstration.
            emptyList<Payment>()
        } catch (ex: Exception) {
            Timber.e(ex, "Error retrieving offline payments.")
            emit(results)
            return@flow
        }

        // If the above line actually returns offline Payment objects, we process them:
        // For demonstration, the repository's syncOfflinePayments might handle them internally.

        // The specification mentions "Handle conflicts" - a placeholder to illustrate:
        // handleConflictsIfAny(queuedPayments)

        // The repository method might have updated the local DB for each payment. Alternatively,
        // we gather results Payment by Payment:
        queuedPayments.forEach { payment ->
            try {
                // Attempt to process the queued payment again
                val processedPayment = paymentRepository.processPayment(payment)
                results.add(Result.success(processedPayment))
                analytics.trackPaymentMetric("payment_synced_success", processedPayment)
            } catch (ex: Exception) {
                Timber.e(ex, "Failed to sync offline payment: ${payment.id}")
                results.add(Result.failure(ex))
                analytics.trackPaymentMetric("payment_synced_failure", payment)
            }
        }

        // STEP 5: Update sync status is presumably done within repository methods or we do it here
        // paymentRepository.updatePaymentStatus(...)

        // STEP 6: Return sync results
        emit(results)
    }

    // ---------------------------------------------------------------------------------------------
    // PRIVATE HELPER: isNetworkAvailable
    // Description: Placeholder function to simulate a network connectivity check. In real usage,
    // you'd integrate with a system service or a network monitoring library to track connectivity.
    // ---------------------------------------------------------------------------------------------
    private fun isNetworkAvailable(): Boolean {
        // Implementation might query the system's connectivity manager or a specialized library
        return true // For demonstration, assume always true or false as needed
    }
}

File: src/ios/DogWalking/Domain/UseCases/PaymentUseCase.swift
//
//  PaymentUseCase.swift
//  DogWalking
//
//  Created by Elite Software Architect on 2023-10-10.
//  This file implements a thread-safe use case for processing payments,
//  handling tips, refunds, and user payment history with comprehensive
//  error handling, fraud detection considerations, and PCI compliance.
//
//  © 2023 DogWalking Inc. All rights reserved.
//

import Foundation // iOS 13.0+ (Core iOS functionality)
import Combine    // iOS 13.0+ (reactive programming support)

// MARK: - Internal Imports (per specification)
import class Payment       // from "../Models/Payment.swift"
import enum PaymentStatus
import class PaymentRepository // from "../../Data/Repositories/PaymentRepository.swift"

// MARK: - Placeholder Import for Payment Monitoring
// In a production environment, this import would reference a real payment monitoring utility.
import class PaymentMonitor // version 1.0 (hypothetical monitoring tool)

/**
 A thread-safe use case class implementing secure payment business logic and orchestrating
 payment operations with comprehensive error handling, monitoring, and PCI compliance.

 This class addresses:
 - Secure payments
 - Automated billing
 - Transaction history retrieval
 - Enhanced security and monitoring
*/
public final class PaymentUseCase {

    // MARK: - Properties

    /// A reference to the payment repository, which handles low-level payment data operations
    /// such as processing, refunds, and fetching payment history.
    private let paymentRepository: PaymentRepository

    /// A lock ensuring that all payment-related operations are thread-safe and
    /// avoid data races in concurrent scenarios.
    private let paymentLock: NSLock

    /// A monitoring tool (hypothetical) used for tracking payment processing times,
    /// fraud detection triggers, and other performance metrics in compliance with
    /// stringent auditing requirements.
    private let monitor: PaymentMonitor

    // MARK: - Initialization

    /**
     Initializes the payment use case with required dependencies, setting up
     monitoring, fraud detection, validation rules, and concurrency controls.

     Steps performed:
     1. Store payment repository reference
     2. Initialize payment lock for thread safety
     3. Configure payment monitoring
     4. Set up fraud detection
     5. Initialize payment validation rules

     - Parameters:
       - paymentRepository: A concrete implementation of PaymentRepository for
         thread-safe payment operations.
       - monitor: A PaymentMonitor instance for tracking and auditing payment flows.
    */
    public init(paymentRepository: PaymentRepository, monitor: PaymentMonitor) {
        self.paymentRepository = paymentRepository
        self.monitor = monitor
        self.paymentLock = NSLock()

        // Additional setup steps such as fraud detection and validation rule configuration
        // would go here in a real implementation. This might include hooking into
        // specialized services or external libraries for advanced risk analysis.
    }

    // MARK: - Public Methods

    /**
     Processes a payment for a completed dog walk, performing validation, fraud checks,
     fee calculations, and repository orchestration with comprehensive monitoring.

     Steps:
     1. Acquire payment lock for thread safety
     2. Validate payment amount and details
     3. Check for suspicious activity patterns
     4. Apply service fees and calculate final amount
     5. Process payment through repository with retry mechanism
     6. Log transaction details for audit
     7. Monitor payment processing time
     8. Release payment lock
     9. Return processed payment or error details

     - Parameter payment: The Payment object to be processed.
     - Returns: A publisher that emits the successfully processed Payment
       or an Error detailing any failures.
    */
    public func processWalkPayment(payment: Payment) -> AnyPublisher<Payment, Error> {
        return Future<Payment, Error> { [weak self] promise in
            guard let strongSelf = self else {
                promise(.failure(NSError(domain: "PaymentUseCase", code: -1, userInfo: [
                    NSLocalizedDescriptionKey: "PaymentUseCase was deallocated"
                ])))
                return
            }

            // 1. Acquire lock
            strongSelf.paymentLock.lock()

            // Use defer to ensure the lock is released even if an error paths out early.
            defer {
                strongSelf.paymentLock.unlock()
            }

            // 2. Validate payment amount and details (uses Payment's internal validation or other checks).
            // Payment already includes validation logic and status transitions, but we can add extra checks if needed.
            // For example, verifying currency usage or max limit again, beyond Payment's creation-time checks.

            // 3. Check for suspicious activity patterns (placeholder).
            // In real usage, we'd do advanced risk analysis or call out to a risk engine.
            // e.g., strongSelf.monitor.checkFraudPatterns(for: payment)

            // 4. Apply service fees or final amount adjustments if needed (placeholder).
            // This step might involve adding a convenience fee or adjusting the final total.
            // For demonstration, we assume payment.amount remains as-is.

            // 5. Begin monitoring the operation.
            strongSelf.monitor.beginOperation(name: "processWalkPayment")

            // 6. Process payment through repository with built-in retry or error handling.
            let repositoryPublisher = strongSelf.paymentRepository.processPayment(payment: payment)
                .handleEvents(receiveOutput: { _ in
                    // 6. Log transaction details for auditing.
                    // Example: You might also log to a secure ledger or monitoring tool here.
                    strongSelf.monitor.logEvent("WalkPaymentProcessingCompleted for paymentId=\(payment.id)")
                }, receiveCompletion: { completion in
                    // 7. Conclude monitoring after repository pipeline finishes.
                    strongSelf.monitor.endOperation(name: "processWalkPayment")
                })
                .eraseToAnyPublisher()

            // 8. Return the repository publisher as our final result (wrapped in the promise).
            let cancellable = repositoryPublisher.sink(
                receiveCompletion: { completion in
                    if case let .failure(error) = completion {
                        promise(.failure(error))
                    }
                },
                receiveValue: { processedPayment in
                    // 9. Return processed payment or error details
                    promise(.success(processedPayment))
                }
            )

            // If we need to store the cancellable for cancellation, we could do so. Here we discard it.
            _ = cancellable
        }
        .eraseToAnyPublisher()
    }

    /**
     Processes a tip payment for a walker, verifying basic validation rules and updating
     the walker’s earnings. It includes concurrency control, logging, and monitoring.

     Steps:
     1. Acquire payment lock
     2. Validate tip amount and limits
     3. Process tip through repository
     4. Update walker earnings atomically (placeholder step, would be integrated with domain logic)
     5. Log tip transaction
     6. Release payment lock
     7. Return tip processing result

     - Parameter payment: A Payment object initialized specifically for the tip transaction.
     - Returns: A publisher emitting the processed Payment or an Error if something fails.
    */
    public func processTipPayment(payment: Payment) -> AnyPublisher<Payment, Error> {
        return Future<Payment, Error> { [weak self] promise in
            guard let strongSelf = self else {
                promise(.failure(NSError(domain: "PaymentUseCase", code: -1, userInfo: [
                    NSLocalizedDescriptionKey: "PaymentUseCase was deallocated"
                ])))
                return
            }

            // 1. Acquire lock
            strongSelf.paymentLock.lock()
            defer {
                strongSelf.paymentLock.unlock()
            }

            // 2. Validate tip amount. We could do extra checks beyond Payment's built-in logic.
            // For example, ensuring tip doesn't exceed some maximum percentage or a certain threshold.

            // 3. Start monitoring tip processing.
            strongSelf.monitor.beginOperation(name: "processTipPayment")

            // 4. Use the repository to actually complete the tip transaction.
            let tipPublisher = strongSelf.paymentRepository.processPayment(payment: payment)
                .flatMap { processedTip -> AnyPublisher<Payment, Error> in
                    // 4A. Update walker earnings (placeholder).
                    // In real usage, we might have to call another service, e.g.:
                    // earningsRepository.updateWalkerEarnings(walkerId: processedTip.walkerId, amount: processedTip.amount)

                    // 5. Log the tip transaction details.
                    strongSelf.monitor.logEvent("TipPaymentProcessed for walker=\(processedTip.walkerId) amount=\(processedTip.amount)")

                    // Return the processed Payment as an immediate publisher.
                    return Just(processedTip)
                        .setFailureType(to: Error.self)
                        .eraseToAnyPublisher()
                }
                .handleEvents(receiveCompletion: { _ in
                    // Close out monitoring on completion
                    strongSelf.monitor.endOperation(name: "processTipPayment")
                })
                .eraseToAnyPublisher()

            let cancellable = tipPublisher.sink(
                receiveCompletion: { completion in
                    if case let .failure(error) = completion {
                        promise(.failure(error))
                    }
                },
                receiveValue: { finalPayment in
                    // 7. Return tip processing result
                    promise(.success(finalPayment))
                }
            )
            _ = cancellable
        }
        .eraseToAnyPublisher()
    }

    /**
     Requests a refund for an already processed payment, ensuring eligibility checks,
     calculating the correct refund amount, and updating payment status with the repository.

     Steps:
     1. Acquire payment lock
     2. Validate refund eligibility
     3. Calculate refund amount including fees
     4. Process refund through repository
     5. Update payment status atomically
     6. Log refund transaction
     7. Release payment lock
     8. Return refund result

     - Parameter paymentId: The unique identifier of the payment to be refunded.
     - Returns: A publisher emitting the updated Payment reflecting the refund status,
       or an Error if something fails in the process.
    */
    public func requestRefund(paymentId: String) -> AnyPublisher<Payment, Error> {
        return Future<Payment, Error> { [weak self] promise in
            guard let strongSelf = self else {
                promise(.failure(NSError(domain: "PaymentUseCase", code: -1, userInfo: [
                    NSLocalizedDescriptionKey: "PaymentUseCase was deallocated"
                ])))
                return
            }

            // 1. Acquire lock
            strongSelf.paymentLock.lock()
            defer {
                strongSelf.paymentLock.unlock()
            }

            // 2. Mark start of refund monitoring
            strongSelf.monitor.beginOperation(name: "requestRefund")

            // 3. Because the repository handles eligibility checks in conjunction with the Payment model’s status,
            //    pass the request to the repository, letting it verify that the Payment has previously completed.
            //    If additional domain logic is needed, it would be placed here (e.g., limit windows for refunds).

            // 4. Process the refund
            let refundPublisher = strongSelf.paymentRepository.refundPayment(paymentId: paymentId, reason: "UserRequestedRefund")
                .handleEvents(receiveOutput: { refundedPayment in
                    // 5. Payment status is updated inside the repository. We can do any additional atomic updates here if needed.
                    // 6. Log the refund transaction
                    strongSelf.monitor.logEvent("RefundRequested for paymentId=\(refundedPayment.id) status=\(refundedPayment.status)")
                }, receiveCompletion: { _ in
                    // 7. Conclude monitoring after repository pipeline finishes
                    strongSelf.monitor.endOperation(name: "requestRefund")
                })
                .eraseToAnyPublisher()

            // 8. Return result
            let cancellable = refundPublisher.sink(
                receiveCompletion: { completion in
                    if case let .failure(error) = completion {
                        promise(.failure(error))
                    }
                },
                receiveValue: { finalRefundedPayment in
                    promise(.success(finalRefundedPayment))
                }
            )
            _ = cancellable
        }
        .eraseToAnyPublisher()
    }

    /**
     Retrieves a paginated payment history for a specific user, applying
     domain logic to validate user access, apply filters, and mask sensitive data.

     Steps:
     1. Validate user access rights (placeholder or external logic)
     2. Apply pagination parameters
     3. Fetch payment history from repository
     4. Filter by payment status (if needed)
     5. Sort by date descending
     6. Apply data masking for sensitive info
     7. Return paginated results

     - Parameters:
       - userId: The identifier of the user for whom to retrieve a payment history.
       - pagination: A struct specifying the pagination parameters (page/pageSize).
     - Returns: A publisher emitting the requested paginated history or an Error.
    */
    public func getUserPaymentHistory(userId: String,
                                     pagination: PaginationParams) -> AnyPublisher<PaginatedPayments, Error> {
        return Future<PaginatedPayments, Error> { [weak self] promise in
            guard let strongSelf = self else {
                promise(.failure(NSError(domain: "PaymentUseCase", code: -1, userInfo: [
                    NSLocalizedDescriptionKey: "PaymentUseCase was deallocated"
                ])))
                return
            }

            // 1. Validate user access rights (placeholder).
            // For instance, ensure the user requesting the history is the owner or has proper roles.

            // 2. Convert PaginationParams to the repository's Pagination struct
            let repoPagination = Pagination(page: pagination.page, pageSize: pagination.pageSize)

            // 3. We begin a monitoring operation for retrieving payment history
            strongSelf.monitor.beginOperation(name: "getUserPaymentHistory")

            // 4. Perform the repository call with some default or domain-specific filtering
            //    This example uses an empty PaymentFilter to fetch everything.
            let historyPublisher = strongSelf.paymentRepository.getPaymentHistory(
                userId: userId,
                filter: PaymentFilter(status: nil, searchTerm: nil),
                pagination: repoPagination
            )
            .map { paginated -> PaginatedPayments in
                // 4. Filter by payment status if needed (omitted if status is nil).
                // 5. Sort by date descending: we can assume repository or code here. Example for demonstration:
                var sortedPayments = paginated.payments.sorted { lhs, rhs in
                    // If Payment records had a date property for ordering, we'd compare them here.
                    // For demonstration, let's use lhs.createdAt vs. rhs.createdAt.
                    lhs.createdAt > rhs.createdAt
                }

                // 6. Apply data masking for sensitive info. Example: removing or redacting
                // partial data from Payment metadata. We'll do a simple pass here:
                for payment in sortedPayments {
                    // You might remove or mask certain metadata fields if they contain PII.
                    // e.g., payment.metadata.removeValue(forKey: "internalPII")
                    _ = payment // placeholder
                }

                // Rebuild PaginatedPayments with the updated array
                return PaginatedPayments(
                    payments: sortedPayments,
                    currentPage: paginated.currentPage,
                    totalItems: paginated.totalItems,
                    pageSize: paginated.pageSize
                )
            }
            .handleEvents(receiveCompletion: { _ in
                // 7. End operation monitoring
                strongSelf.monitor.endOperation(name: "getUserPaymentHistory")
            })
            .eraseToAnyPublisher()

            let cancellable = historyPublisher.sink(
                receiveCompletion: { completion in
                    if case let .failure(error) = completion {
                        promise(.failure(error))
                    }
                },
                receiveValue: { updatedPaginated in
                    // 8. Return the final results
                    promise(.success(updatedPaginated))
                }
            )
            _ = cancellable
        }
        .eraseToAnyPublisher()
    }
}

/**
 A placeholder struct that represents the pagination parameters needed
 by the PaymentUseCase. In a real scenario, this would likely be shared
 across multiple use cases or bound to UI input, then mapped to the
 PaymentRepository's Pagination struct.
 */
public struct PaginationParams {
    /// The zero-based page index requested.
    public let page: Int
    /// The number of items to be returned per page.
    public let pageSize: Int

    public init(page: Int, pageSize: Int) {
        self.page = page
        self.pageSize = pageSize
    }
}